src/main/scala/org/broadinstitute/dsde/agora/server/webservice/PerRequest2.scala

```diff
@@ -0,0 +1,40 @@
+package org.broadinstitute.dsde.agora.server.webservice
+
+import scala.language.postfixOps
+import akka.http.scaladsl.marshalling.{Marshaller, ToResponseMarshaller}
+import akka.http.scaladsl.model.HttpHeader
+
+import scala.concurrent.ExecutionContext
+
+object PerRequest2 {
+  implicit def requestCompleteMarshaller(implicit ec: ExecutionContext): ToResponseMarshaller[PerRequestMessage] =
+    Marshaller {
+      _: ExecutionContext => {
+        case requestComplete: RequestComplete[_] =>
+          requestComplete.marshaller(requestComplete.response)
+        case requestComplete: RequestCompleteWithHeaders[_] =>
+          requestComplete.marshaller(requestComplete.response).map(_.map(_.map(_.mapHeaders(_ ++ requestComplete.headers))))
+      }
+    }
+
+  sealed trait PerRequestMessage
+  /**
+    * Report complete, follows same pattern as spray.routing.RequestContext.complete; examples of how to call
+    * that method should apply here too. E.g. even though this method has only one parameter, it can be called
+    * with 2 where the first is a StatusCode: RequestComplete(StatusCode.Created, response)
+    */
+  case class RequestComplete[T](response: T)(implicit val marshaller: ToResponseMarshaller[T]) extends PerRequestMessage
+
+  /**
+    * Report complete with response headers. To response with a special status code the first parameter can be a
+    * tuple where the first element is StatusCode: RequestCompleteWithHeaders((StatusCode.Created, results), header).
+    * Note that this is here so that RequestComplete above can behave like spray.routing.RequestContext.complete.
+    */
+  case class RequestCompleteWithHeaders[T](response: T, headers: HttpHeader*)(implicit val marshaller: ToResponseMarshaller[T]) extends PerRequestMessage
+
+  /**
+    * Report complete with a path to use as a Location header
+    */
+  case class RequestCompleteWithLocation[T](response: T, location: String)(implicit val marshaller: ToResponseMarshaller[T]) extends PerRequestMessage
+
+}
```

Comment:
I'm questioning the value of `PerRequest2`. Does this abstraction offer any positive value, as opposed to routes completing on their own? The abstraction:
```
complete(RequestComplete(httpStatus, statusCheckResponse))
```
looks more confusing to me than the direct approach of:
```
complete(httpStatus, statusCheckResponse)
```

on top of that, with so many developers used to seeing `RequestComplete` as a spray utility, I think it's hard to read.

What is the value-add? Is it carrying the marshaller along to the point of the response? It feels like there could be another solution to that.

Reply:
I was thinking about the same thing last night, I couldn't really put it into words so I made a PR against this branch

https://github.com/broadinstitute/agora/pull/239

Reply:
FWIW the party line from Lightbend was to use this opportunity to break free from the per request model altogether. It adds a lot of complexity (doubly so now) for what turns out to be little to no gain in most cases

Reply:
No good reason other than minimizing code changes which may reduce the risk of regressions. At the time, having @dvoet's Rawls counterpart to follow also gave me confidence. Rawls probably has more reasons to go that way than Agora, though given the amount of APIs it has. 

I'd be happy to get rid of `PerRequest` similarly to the way @aednichols proposed. If it makes sense to do it in two phases (first finish the conversion the current way and then get rid of `PerRequest`) to be safe, I can also go that way. Any thoughts?

Reply:
I don't see benefit to a two-phase conversion (1. via a PerRequest stub; 2. "standard" akka-http). I'd rather just convert it to the way we want it instead of dragging it out. Thanks!

Reply:
Before really diving in, I had also thought (and suggested to Kaan) that two phases might be safer, but I see everyone's point